File: app/src/main/java/com/example/avjindersinghsekhon/minimaltodo/SettingsFragment.java
package com.example.avjindersinghsekhon.minimaltodo;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.PreferenceFragment;

import com.example.avjindersinghsekhon.minimaltodo.contract.Contract;
import com.example.avjindersinghsekhon.minimaltodo.view.main.MainActivity;

public class SettingsFragment extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener{

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.preferences_layout);
   }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        PreferenceKeys preferenceKeys = new PreferenceKeys(getResources());
        if(key.equals(preferenceKeys.night_mode_pref_key)){
            SharedPreferences themePreferences = getActivity().getSharedPreferences(Contract.THEME_PREFERENCES, Context.MODE_PRIVATE);
            SharedPreferences.Editor themeEditor = themePreferences.edit();
            //We tell our MainLayout to recreate itself because mode has changed
            themeEditor.putBoolean(Contract.RECREATE_ACTIVITY, true);

            CheckBoxPreference checkBoxPreference = (CheckBoxPreference)findPreference(preferenceKeys.night_mode_pref_key);
            if(checkBoxPreference.isChecked()){
                //Comment out this line if not using Google Analytics
                themeEditor.putString(Contract.THEME_SAVED, Contract.LIGHT_THEME);
            }
            else{
                themeEditor.putString(Contract.THEME_SAVED, Contract.LIGHT_THEME);
            }
            themeEditor.apply();

            getActivity().recreate();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }
}

File: app/src/main/java/com/example/avjindersinghsekhon/minimaltodo/SettingsFragment.kt
package com.example.avjindersinghsekhon.minimaltodo

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.CheckBoxPreference
import android.preference.PreferenceFragment

/**
 * Created by hardyeats on 2017-07-22.
 */

class SettingsFragment : PreferenceFragment(), SharedPreferences.OnSharedPreferenceChangeListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        addPreferencesFromResource(R.xml.preferences_layout)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        val preferenceKeys = PreferenceKeys(resources)
        if(key == preferenceKeys.night_mode_pref_key) {
            val themePreferences = activity.getSharedPreferences(MainActivity.THEME_PREFERENCES, Context.MODE_PRIVATE)
            themePreferences.edit().apply {
                //We tell our MainLayout to recreate itself because mode has changed
                putBoolean(MainActivity.RECREATE_ACTIVITY, true)
                val checkBoxPreference  = findPreference(preferenceKeys.night_mode_pref_key) as CheckBoxPreference
                if(checkBoxPreference.isChecked)
                    putString(MainActivity.THEME_SAVED, MainActivity.DARKTHEME)
                else
                    putString(MainActivity.THEME_SAVED, MainActivity.LIGHTTHEME)
                apply()
            }
            activity.recreate()
        }
    }

    override fun onResume() {
        super.onResume()
        preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }
}